doped/utils/displacements.py

```diff
@@ -553,3 +489,726 @@ def _plotly_plot_total_disp(
+        bulk_sc_index = mappings_dict[i]  # Map to bulk sc
+        bulk_site = bulk_sc[bulk_sc_index]  # Get site in bulk sc
+        # Calculate displacement (need to account for pbc!)
+        # First final point, then initial point
+        frac_disp = pbc_diff(site.frac_coords, bulk_site.frac_coords)  # in fractional coords
+        disp = bulk_sc.lattice.get_cartesian_coords(frac_disp)  # in Angstroms
+        # Distance to defect site (last site in defect sc)
+        distance = defect_sc_with_site.get_distance(i, defect_site_index)  # len(defect_sc_with_site) - 1)
+
+        disp_dict["Index (defect)"].append(i)
+        disp_dict["Displacement"].append(disp)
+        disp_dict["Displacement Norm"].append(np.linalg.norm(disp, ord=2))
+        disp_dict["Distance to defect"].append(distance)
+        disp_dict["Species_with_index"].append(f"{site.specie.name}({i})")
+        disp_dict["Species"].append(site.specie.name)
+        disp_dict["X sites in cartesian coordinate (defect)"].append(
+            _shift_defect_site_to_center_of_the_supercell(
+                site.frac_coords, sc_defect_frac_coords, bulk_sc
+            )[0]
+        )
+        disp_dict["Y sites in cartesian coordinate (defect)"].append(
+            _shift_defect_site_to_center_of_the_supercell(
+                site.frac_coords, sc_defect_frac_coords, bulk_sc
+            )[1]
+        )
+        disp_dict["Z sites in cartesian coordinate (defect)"].append(
+            _shift_defect_site_to_center_of_the_supercell(
+                site.frac_coords, sc_defect_frac_coords, bulk_sc
+            )[2]
+        )
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", "invalid value encountered in scalar divide")
+
+    # sort each list in disp dict by index of species in bulk element list, then by distance to defect:
+    element_list = [
+        el.symbol for el in defect_entry.defect.structure.composition.elements
+    ]  # host elements
+    element_list += sorted(
+        [  # extrinsic elements, sorted alphabetically for deterministic ordering in output:
+            el.symbol
+            for el in defect_entry.defect.defect_structure.composition.elements
+            if el.symbol not in element_list
+        ]
+    )
+    # Combine the lists into a list of tuples, then sort, then unpack:
+    combined = list(zip(*disp_dict.values()))
+    combined.sort(
+        key=lambda x: (element_list.index(x[1]), x[4], x[0])
+    )  # Sort by species, then distance, then index
+
+    (
+        disp_dict["Index (defect)"],
+        disp_dict["Species"],
+        disp_dict["Species_with_index"],
+        disp_dict["Displacement"],
+        disp_dict["Displacement Norm"],
+        disp_dict["Distance to defect"],
+        disp_dict["X sites in cartesian coordinate (defect)"],
+        disp_dict["Y sites in cartesian coordinate (defect)"],
+        disp_dict["Z sites in cartesian coordinate (defect)"],
+    ) = zip(*combined)
```

Comment:
A good chunk of this code is duplicated from the `calc_site_displacements` function. It would be best for code efficiency and maintenance if this is pulled out into a helper function that then is used in both functions (with flags for `relative_to_defect`, `vector_to_project_on` and `ellipsoid` to determine which `disp_dict` data is needed?).
Could you implement this please?